### DISCUSSION TITLE
Handling exceptions from model update

### DISCUSSION THREAD

**Discussion by igormorgado** [collerek/ormar#421] [Q&A] [answered]
If I try to update a model with `MODEL.update()` or save it with  `MODEL.save()`, when it fails for violate a unique constraint it raises the exception `asyncpg.exceptions.UniqueViolationError`, obviously that happens because I'm using `Postgres`, but catching this exception will make my code coupled with asyncpg and postgres, hence, defeating some of the advantages of using a ORM.

Is possible to catch it at Ormar level?  Like Ormar.exception.ViolationError?

By the way, that also happens in other parts ormar usage, for example when violating some validation error it raises a hard to process error from pydantic `pydantic.error_wrappers.ValidationError`..

**Comment by collerek** [comment, marked as answer]
Yes I was thinking about that too, did something similar in fastapi-crudrouter as in that case you need to check if user have separate drivers installed or not (for MySQL, postgres etc.). That might appear but would be a breaking change so at soonest in 0.11.

**Comment by 5h44n** [comment]
Would love to help out on this feature. Is there ongoing work on it (happy to start if not), or is it being delayed until 0.11? Not sure what the roadmap looks like. 

Based off @igormorgado's comment, was thinking the following approach:

for validation errors at the pydantic level: 
- implement a `pydantic_validation.py` in `ormar.models.helpers` that attempts to instantiate the respective pydantic model with the input data in creates/updates and raises a new `ormar.exceptions.ModelValidationError`.

for db integrity errors:
- implement something similar to `fastapi-crudrouter/ormar.py`, ultimately raising the same new `ormar.exceptions.ModelValidationError` exception from above to the user

dont have an approach off the top for bulk creates/updates since those might bypass pydantic, but maybe it would work the same as catching any db integrity errors in a single update/create

**Reply by collerek** [reply]
Hi,

That is not so easy, unfortunately.

In crud router you have only one error -> IntegrityError, while in database usage you have plenty of them. 
Should we catch only this one and the rest should be as is? Why only this one?

Moreover, crud router supports only 3 drivers, while databases support already 5 of them. That also creates the dependency between ormar and the drivers themselves, not only the databases, as you need to catch specific exceptions from each driver so you need to import them (and wrap them in tries if the user does not have this driver installed, etc.). 

If we decide to include more exceptions, after all, you might want to build different logic to solve IntegrityErrors, ForeignKeyViolations, IncorrectData, or ConnectionError so you cannot simply group them under one ormar Exception (you limit already existing functionality and it's a bad practice). That would have to be a hierarchy of exceptions that needs to be maintained based on 5 separate drivers.

As for pydantic Validation errors, the problem is that fastapi automatically intercepts pydantic errors and presents you with a nice response with proper error code, etc. It's not working out of the box for other exceptions types - you need to write and register a custom error handler. That's why I am against wrapping pydantic errors - it would break every ormar-fastapi application that is already working.
